Adds get leaderboard stats api

arunava59 opened the pull request:
closes https://github.com/coronagoapp/coronago-backend/issues/4

**Why**
The user wants to get the leaderboard information - which portrays top 10 users sorted descending by wallet balance, and also the rank and balance of the logged in user as well.

**What**

- Adds a new API endpoint to fetch the above information.
- Makes two queries, one to fetch global leaderboard stats, the other to fetch current user rank.
- Adds a random user generator, it generates a million users with random wallet balances.
- Adds Rspec examples for testing
- Removes few dependencies that are not used
  - `require capybara/rspec` in `rails_helper.rb`
  - `require helpers` in `spec_helper.rb`

When current user is not in Top 10 list
![image](https://user-images.githubusercontent.com/62326676/77878536-232bc580-7276-11ea-86fa-118f2c33abad.png)


When current user is in Top 10 list
![image](https://user-images.githubusercontent.com/62326676/77878536-232bc580-7276-11ea-86fa-118f2c33abad.png)


/cc @prasadpilla @nronas

prasadpilla commented:
Hey @arunava59, 

1. Can we performance test this for at least a million records? I understand that this is a quick solution, but we want to make sure it works for at least 1M records so that we'd have reaction time time to build more scalable solution.
2. Did you test this in a scenario where the current user is not in the top rank?
3. Can you use a PR template with more details like [this one](https://github.com/coronagoapp/coronago-backend/pull/13)

arunava59 commented:
Hey @prasadpilla ,

1. I added a rake task that generates a million users with random balances. Tested the api on it, took about 2.1 seconds.
2. Yes I have tested it in both cases, it is working well. Please find the screenshots in the PR description.
3. I have updated the PR template. Please check.

prasadpilla commented:
> Hey @prasadpilla ,
> 
> 1. I added a rake task that generates a million users with random balances. Tested the api on it, took about 2.1 seconds.
> 2. Yes I have tested it in both cases, it is working well. Please find the screenshots in the PR description.
> 3. I have updated the PR template. Please check.

@arunava59 Great work, Awesome!!!

prasadpilla commented on lib/tasks/generate_random_data.rake:
```diff
@@ -0,0 +1,26 @@
+namespace :generate_random_data do
```
A bit more specific name would be great i guess? something like "generate_users_data"

arunava59 replied:
I had the intention of using this generic namespace for populating tables with random data as per requirement. The users table is just one of the many, and so if required at some later stage we can add more tasks in this file to populate other tables as well.
Do you want me to make this file specific to users only ?

prasadpilla replied:
@arunava59 oh yeah, great then.

prasadpilla commented on lib/tasks/generate_random_data.rake:
```diff
@@ -0,0 +1,26 @@
+namespace :generate_random_data do
+
+  desc 'Generates a million random entries in users table'
```
Can you make the number configurable, move it up into a constant and probably parse from command line args? 

Default can be 1M

arunava59 replied:
Yes sure, doing that ASAP.

arunava59 replied:
Done!

prasadpilla commented:
@arunava59 Can you also add few units tests since this is one of the core features?

arunava59 commented:
Okay, would add the unit tests soon.

nronas commented on config/routes.rb:
```diff
@@ -52,6 +52,8 @@
         end
       end
+
+      get 'leaderboard_stats' => 'users#fetch_leader_board_stats'
```
Since this is already a custom resource shall we resource-fy like:

```suggestion
      resource :leaderboard, only: show
```

nronas commented on app/models/user.rb:
```diff
@@ -75,6 +78,28 @@ def remove_devise_validations
     end
   end
 
+  def self.fetch_leader_board_stats(current_user)
```
Just for posterity:

I think two faster queries with no temp space allocation here will be more efficient:

```ruby
top_ten = User.order('wallet_balance desc').limit(10)
current_rank = User.count('wallet_balance > ?', current_user.wallet_balance)

{top_ten: top_den, current_rank: current_rank}
```

nronas commented:
Lets also document the new endpoint in postman https://documenter.getpostman.com/view/494437/SzYW1z2B

nronas commented on app/controllers/api/v1/leaderboards_controller.rb:
```diff
@@ -0,0 +1,16 @@
+  private
+
+  def set_user
+    @user = User.find_by!(id: params[:id])
```
Since the show method is a singular resource, does not have a param[:id] passed by. 

```
api_v1_leaderboard GET    /api/v1/leaderboard(.:format)                                                            api/v1/leaderboards#show
```

If i understand this correctly, we want to get the current user here, which we already have it from the current_user helper method: https://github.com/coronagoapp/coronago-backend/blob/master/app/controllers/api/base_controller.rb#L13

arunava59 replied:
Initially I wanted to use the helper method like you mentioned, but one look at the users_controller made me realize that we are setting user, and then checking ownership. Thus we are maintaining two instance variables namely: @user and @api_current_user.
Should I use the helper method instead ?

nronas replied:
yeah lets use the helper!

nronas commented on app/models/user.rb:
```diff
@@ -75,6 +78,33 @@ def remove_devise_validations
+    return {
+        top_users: top_users_json,
+        user: current_user_json
+    }
```
We want even if the current user is in the top ten list to populate is ranking on the current_user_json
```suggestion
  essential_fields  = [:id, :name, :profile_picture_url, :country_code, :wallet_balance]

  # Fetches top users with maximum wallet_balance.
  top_users = User.order('wallet_balance desc').limit(LEADERBOARD_LIMIT)

  # Checks if current user is in top users
  # If yes, then we are done.
  # Else, finds rank of current user and build its json
  current_user_json = current_user.as_json(only: essential_fields)

  current_user_json['wb_rank'] = if top_users.include?(current_user)
                                   top_users.index(current_user) + 1
                                 else
                                   User.where('wallet_balance > ?', current_user.wallet_balance).count + 1
                                 end

  top_users_json.each_with_index do |user, idx|
    user['wb_rank'] = idx + 1
  end

  {
    top_users: top_users.as_json(essential_fields),
    user: current_user_json
  }
```

arunava59 replied:
In the suggested code if the current user is already in the top user list, won't we be calling `as_json` twice for this user instance : one while performing `current_user.as_json` and the other while performing `top_users.as_json` ?
I think you are concerned that if the current user is already in the top still the variable `current_user_json` needs to be populated, but I think it is already happening in the existing code. Please check ?

arunava59 commented:
Have done the necessary changes. Please check.
